fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/MacroFunctionValue.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * MacroFunctionValue.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2024 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades.values;
+
+import com.apple.foundationdb.record.EvaluationContext;
+import com.apple.foundationdb.record.PlanDeserializer;
+import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.PlanSerializationContext;
+import com.apple.foundationdb.record.planprotos.PMacroFunctionValue;
+import com.apple.foundationdb.record.planprotos.PValue;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStoreBase;
+import com.apple.foundationdb.record.query.plan.cascades.SemanticException;
+import com.apple.foundationdb.record.query.plan.cascades.typing.Type;
+import com.apple.foundationdb.record.query.plan.cascades.typing.Typed;
+import com.apple.foundationdb.record.query.plan.cascades.values.translation.TranslationMap;
+import com.google.auto.service.AutoService;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * The class defines a function.
+ * The function takes in a list of arguments as QuantifiedObjectValue, of which alias = uniqueId, resultType = argument type.
+ * and a body value, executing the function is by replacing the QuantifiedObjectValue in the body with the actual argument.
+ */
+public class MacroFunctionValue extends AbstractValue {
+
+    @Nonnull
+    private final List<Value> argList;
```

Comment:
The style is to name a list of(unless ambiguous) just the plural of the things the list holds and to add the type of it to the name (in the planner). Also, avoid contractions and abbreviations. So this one should be something like `argumentValues`

Reply:
this one becomes "parameterTypes" (same as BuiltInFunction), and "parameterIdentifiers" in the new MacroFunction class.